test/extensions/filters/listener/common/fuzz/listener_filter_fuzzer.cc

```diff
@@ -10,19 +8,87 @@ void ListenerFilterFuzzer::fuzz(
   try {
-    fuzzerSetup(input);
+    socket_.setLocalAddress(Network::Utility::resolveUrl(input.sock().local_address()));
+  } catch (const EnvoyException& e) {
+    // Socket's local address will be nullptr by default if fuzzed local address is malformed
+    // or missing - local address field in proto is optional
+  }
+  try {
+    socket_.setRemoteAddress(Network::Utility::resolveUrl(input.sock().remote_address()));
   } catch (const EnvoyException& e) {
-    ENVOY_LOG_MISC(debug, "EnvoyException: {}", e.what());
-    return;
+    // Socket's remote address will be nullptr by default if fuzzed remote address is malformed
+    // or missing - remote address field in proto is optional
+  }
+
+  const int nreads = input.data_size(); // Number of reads
+
+  if (nreads > 0) {
+    EXPECT_CALL(socket_, detectedTransportProtocol()).WillRepeatedly(testing::Return("raw_buffer"));
+
+    EXPECT_CALL(os_sys_calls_, recv(FAKE_SOCKET_FD, _, _, MSG_PEEK))
+        .WillOnce(testing::Return(Api::SysCallSizeResult{static_cast<ssize_t>(0), 0}));
+
+    EXPECT_CALL(dispatcher_,
+                createFileEvent_(_, _, Event::FileTriggerType::Edge,
+                                 Event::FileReadyType::Read | Event::FileReadyType::Closed))
+        .WillOnce(testing::DoAll(testing::SaveArg<1>(&file_event_callback_),
+                                 testing::ReturnNew<NiceMock<Event::MockFileEvent>>()));
   }
 
   filter.onAccept(cb_);
-}
 
-void ListenerFilterFuzzer::socketSetup(
-    const test::extensions::filters::listener::FilterFuzzTestCase& input) {
-  socket_.setLocalAddress(Network::Utility::resolveUrl(input.sock().local_address()));
-  socket_.setRemoteAddress(Network::Utility::resolveUrl(input.sock().remote_address()));
+  if (nreads > 0) {
+    // Construct header from single or multiple reads
+    std::string data = "";
+    std::vector<size_t> indices; // Ending indices for each read
+
+    size_t curr = 0;
+
+    for (int i = 0; i < nreads; i++) {
+      data += input.data(i);
+      curr += input.data(i).size();
+      indices.push_back(curr);
+    }
+
+    absl::string_view header(data);
+
+    int nread = 0; // Counter of current read
+
+    {
+      testing::InSequence s;
+
+      if (nreads > 1) {
+        EXPECT_CALL(os_sys_calls_, recv(FAKE_SOCKET_FD, _, _, MSG_PEEK))
+            .WillOnce(testing::InvokeWithoutArgs([]() {
+              return Api::SysCallSizeResult{ssize_t(-1), SOCKET_ERROR_AGAIN};
+            }));
+      }
+
+      EXPECT_CALL(os_sys_calls_, recv(FAKE_SOCKET_FD, _, _, MSG_PEEK))
+          .Times(testing::AnyNumber())
+          .WillRepeatedly(Invoke([&header, &indices, &nread](os_fd_t, void* buffer, size_t length,
+                                                             int) -> Api::SysCallSizeResult {
+            ASSERT(length >= indices[nread]);
+            memcpy(buffer, header.data(), indices[nread]);
+            return Api::SysCallSizeResult{ssize_t(indices[nread++]), 0};
+          }));
+    }
```

Comment:
This seems like something that could usefully be encapsulated into a small utility class. The constructor would set up `data` and the indices, which would be private data members, and you could have a method `.next(void*)` or something that would copy the next prefix into buffer and return the number of bytes written. That keeps all the EXPECT_CALLs here where they belong, and separates out a useful abstraction. You'll probably want another method that returns whether all reads have completed to replace the `nread >= nreads` bit below.

Reply:
Updated in the latest commit!